src/cursor/cur_index.c

```diff
@@ -355,18 +400,53 @@ __curindex_search_near(WT_CURSOR *cursor, int *exact)
 static int
 __curindex_bound(WT_CURSOR *cursor, const char *config)
 {
+    WT_CONFIG_ITEM cval;
     WT_CURSOR *child;
     WT_CURSOR_INDEX *cindex;
     WT_DECL_RET;
     WT_SESSION_IMPL *session;
+    bool inclusive, increment;
 
-    cindex = (WT_CURSOR_INDEX *)cursor;
-    JOINABLE_CURSOR_API_CALL(cursor, session, bound, NULL);
-
-    /* Grab the primary cursor and call bound function. */
     cindex = (WT_CURSOR_INDEX *)cursor;
     child = cindex->child;
+    inclusive = false;
+    increment = false;
+
+    JOINABLE_CURSOR_API_CALL_CONF(cursor, session, bound, config, cfg, NULL);
+    WT_ERR(__wt_config_gets(session, cfg, "action", &cval));
+
+    /* When setting bounds, we need to check that the key is set. */
+    if (WT_STRING_MATCH("set", cval.str, cval.len)) {
+        WT_ERR(__cursor_checkkey(cursor));
+
+        /* Point the public cursor to the key in the child. */
+        __wt_cursor_set_raw_key(child, &cursor->key);
+
+        WT_ERR(__wt_config_gets(session, cfg, "inclusive", &cval));
+        inclusive = cval.val != 0;
+
+        /* Check if we have set the lower bound or upper bound. */
+        WT_ERR(__wt_config_gets(session, cfg, "bound", &cval));
+    }
+
     WT_ERR(child->bound(child, config));
+
+    /*
+     * Index tables internally combines the user chosen columns with the key format of the table to
+     * maintain uniqueness between each key. However user's are not aware of the combining the key
+     * format and cannot set bounds based on the combined index format. Therefore WiredTiger needs
+     * to internally fix this by incrementing one bit to the array in two cases:
+     *  1. If the set bound is lower and it is not inclusive.
+     *  2. If the set bound is upper and it is inclusive.
+     */
+    if (WT_STRING_MATCH("lower", cval.str, cval.len) && !inclusive)
+        WT_ERR(__increment_bound_array(&child->lower_bound, &increment));
```

Comment:
If the `increment_bound_array` will overflow, that is `if (!increment)`, we'll need to do something to ensure that no values are returned.  If main table has formats "r", "2u" and has one entry: 1 -> 0xffff .  Then index will be key format "2ur", and if caller sets lower bound to 0xffff and exclusive, they expect to receive nothing, in fact they will see the entry.  I don't know if there's a special flag on the cursor, or if you can internally set an upper bound to 0x0000.

Reply:
Talked with @ddanderson on what to do here, and decided to add a return for EINVAL. I tried to replicate the problem where we have a max buffer for 2u case but this was the output i got 
```
(gdb) p (char*)child->lower_bound
$6 = 0xca8720 "\202\377\377"
```